### PR TITLE
187384622/new email model - PR1 - Change Email and Make Sure original tests didn't break

### PR DESCRIPTION
**Pull request by perryzjc**
### [Pivotal Tracker Link1][tracker1]

[tracker1]: https://www.pivotaltracker.com/story/show/187384622

## What This PR Does

This pull request is the first in a series intended to facilitate easier code review. It restructures the `teacher` model to enable a `teacher` to have multiple `email_addresses`. The primary goals of this feature are:
- To allow administrators to manually link several email addresses to a teacher, while a teacher can only modify their primary email (typically a school email).
- To split the implementation across three PRs for clarity and manageability. This initial PR focuses on backend model changes, updates to value access in the frontend template and controller, modifications to existing tests, and adjustments to some seed data and fixtures.

Notably, this PR does not introduce new tests. Its primary objective is to ensure that existing tests continue to pass without issues. Subsequent PRs will include:
- A library task script for migrating current data (not schema) to align with the updated data model, accompanied by an RSpec test.
- Frontend enhancements allowing administrators to add multiple personal emails for a teacher and displaying these emails for both administrators and teachers, complete with comprehensive testing.

### Screenshots and Visuals

After running `rails db:migrate` and `rails db:seed`, the following changes are observable:
1. Teacher names and other associated attributes remain visible, except for the `email` field for original entries in the database. This is due to the migration of the database schema without corresponding data migration, leaving the new email model without the original email information. A forthcoming PR will address this with a data migration script.
2. The current frontend does not yet support displaying multiple email addresses and instead shows a primary email alongside an array of non-primary emails for each teacher.
image
image

#### Who authored this PR?

1. Perry (Jingchao) Zhong (@perryzjc)
4. Michael Tao (@realmichaeltao )

### How should this PR be tested?
At least make sure every original feature doesn't break.

* Is there a deploy we can view?
* What do the specs/features test?
Specs & Features are basically same as the original tests, with changes only relevant to some keywords changing & the way mocking the data (such as originally we can directly create a teacher with a email, now we need to associate them together)
* Are there edge cases to watch out for?
For this PR, at least make sure it  didn't break any existing test.

#### Are there any complications to deploying this?

1. `rails db:migrate`
2. `rails db:seed`

### Checklist:

- [ ] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

### CODE REVIEW

**Comment by razztech** [comment]
The backend code looks good for creating the email model and updating the tests and switching everything to use that model. I also like that you are breaking up the task into smaller PRs since this is a big feature that requires a lot of changes to the codebase. I have a couple of things that I would change:
1. I would name the model `email` instead of `email_address` for the sake of simplicity and it just looks better in my opinion
2. I would remove the frontend where you actually display the emails of each user in Ruby array format. You broke up the task into two PRs (backend and model creation + frontend/testing/ux). Thus, this PR should just update the backend and shouldn't change what the user sees on the website. I think that this is the best approach since you want the PRs to be fully independent from each other. You want the app to function properly regardless if the frontend of the feature is implemented or not. That being said, for the example below I would just display the primary email (ball@berkeley.edu) and remove the array of emails from being displayed. 
Screenshot 2024-04-08 at 4 10 31 PM

**Review by fp456 — APPROVED** [review]
Looks good! The code was written concisely and I don't see any major issues except what Razvan mentioned. The testing also seems cohesive and the new lines were properly tested.

Thanks for reducing the PR size. 🙌

**Comment by perryzjc** [comment]
@razztech @fp456, I just updated this PR based on the feedback from you and Michael.

1. Regarding the naming of the email model, Michael prefers it to remain as `email_addresses`, so this part is unchanged.
2. Regarding the personal email display issue, I have now removed personal emails from the table. For the views that originally used personal emails, I am currently displaying the first one (`.first`). In this way, we can ensure that it passes all the existing tests and that customers use it the same way.
3. I also resolved the issue where many primary emails disappeared from the original records. This was because one of my migrations dropped the `email` column from `teacher`. It should not have been done at this time, but rather later, after my second PR regarding data migration.

Below are screenshots of the current effect. Customers can see the views the same as the original, and new teachers can be created using the new backend (with new email model).
image
image
image
image

File: db/migrate/20240407190132_adapt_teachers_after_new_email_addresses.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class AdaptTeachersAfterNewEmailAddresses < ActiveRecord::Migration[6.1]
+  def change
+    # Remove indexes related to email and personal_email
```
Comment (cycomachead) [inline db/migrate/20240407190132_adapt_teachers_after_new_email_addresses.rb]: This comment isn't necessary. 

A better version of this comment might be
"Remove indexes before removing columns to prevent pg errors"
...though I am not sure if this is strictly necessary or correct, so I'd just remove the comment, since deleting the index in this case is harmless.

File: app/models/email_address.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#
+#  fk_rails_...  (teacher_id => teachers.id)
+#
+class EmailAddress < ApplicationRecord
+  belongs_to :teacher
+
+  validates :email, presence: true, uniqueness: true
+  validate :only_one_primary_email_per_teacher
+
+  before_save :downcase_email
```
Comment (cycomachead) [inline app/models/email_address.rb:31]: Let's call this normalize_email

File: app/models/email_address.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  private
+  def only_one_primary_email_per_teacher
+    if primary? && EmailAddress.where(teacher_id:, primary: true).where.not(id:).exists?
+      errors.add(:primary, "There can only be one primary email per teacher.")
+    end
+  end
+
+  def downcase_email
+    self.email = email.downcase
```
Comment (cycomachead) [inline app/models/email_address.rb:42]: Add a `.strip` to remove whitespace at each end. :)

File: app/controllers/teachers_controller.rb [hunk 4]
```diff
@@ -260,4 +280,33 @@ def sanitize_params
   def load_pages
     @pages ||= Page.where(viewer_permissions: Page.viewable_pages(current_user))
   end
+
+  def update_primary_email(primary_email)
+    return unless primary_email.present?
+
+    # First, ensure the current primary email is marked as not primary if it's not the same as the new one
+    @teacher.email_addresses.where(primary: true).each do |email_record|
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:288]: This seems like we should be able to write a `@teacher.primary_email` call since there should only be 1 primary email for each teacher. Not sure why we need a loop here.

File: app/controllers/teachers_controller.rb [hunk 5]
```diff
@@ -95,7 +101,18 @@ def edit
   def update
     load_school
     ordered_schools
-    @teacher.assign_attributes(teacher_params)
+
+    personal_emails_params = params[:teacher].keys.select { |key| key.to_s.start_with?("personal_email") }
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:105]: I think it looks like we should probably separate the form email fields. The parameters logic is pretty complicated here.

File: app/controllers/teachers_controller.rb [hunk 6]
```diff
@@ -95,7 +101,18 @@ def edit
...
+
+    primary_email = params[:teacher].delete(:primary_email)
+    personal_emails = params[:teacher].extract!(*personal_emails_params).values
+
+    # Now, `params[:teacher]` does not contain primary_email or any personal_emailX fields
+    teacher_attr = teacher_params
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:111]: I would just use `teacher_params` directly in the next line.

**Review by cycomachead — CHANGES_REQUESTED** [review]
This is good! A lot here. 

One high level comment: I think perhaps teacher#create and teacher#update should only ever update the primary email. Updating of personal emails should probably happen with separate controller actions. This will make the params / other logic a lot easier. 

That will simplify the forms, controller actions, and params handling.

File: app/models/teacher.rb [hunk 7]
```diff
@@ -292,21 +284,21 @@ def self.csv_export
...
+    # Default return primary email
+    primary_email
+  end
 
-    teacher_personal_email = Teacher.where.not(id: self.id).exists?(["email = :value OR personal_email = :value", { value: self.personal_email }])
-    if teacher_personal_email
-      errors.add(:personal_email, "must be unique across both email and personal_email columns")
-    end
+  def primary_email
+    email_addresses.find_by(primary: true)&.email
```
[the remark this answers is not in the document]
Comment (perryzjc) [reply app/models/teacher.rb]: It looks like this line was reviewed on an outdated code. My revised code since last time was `email_addresses.find_by(primary: true)&.email || self[:email]`. Here I use `self[:email]` instead of `email` directly due to a naming conflicting (`email` is used as a method name for some other reasons.)

File: app/models/teacher.rb [hunk 8]
```diff
@@ -292,21 +284,21 @@ def self.csv_export
...
+  end
+
+  def personal_emails
+    non_primary_emails
+  end
+
+  private
+  def non_primary_emails
+    email_addresses.where(primary: false)&.pluck(:email)
```
[the remark this answers is not in the document]
Comment (perryzjc) [reply app/models/teacher.rb]: Same as my above comment. It looks like this review was on an outdated code. My current one is `email_addresses.where(primary: false)&.pluck(:email)&.first || personal_email`. May be it will be more readable by switching the order to be like `personal_email || email_addresses.where(primary: false)&.pluck(:email)&.first`

File: db/migrate/20240407190132_adapt_teachers_after_new_email_addresses.rb (hunk 1 above)
Comment (perryzjc) [reply db/migrate/20240407190132_adapt_teachers_after_new_email_addresses.rb]: This review was also based on outdated code. Prior to this review, I had pushed an update that simply removed this migration for this pull request. This was because it dropped the column, resulting in the loss of data (relevant to email), which should not have been done until the data migration script was executed.

File: app/controllers/teachers_controller.rb [hunk 9]
```diff
@@ -42,22 +42,25 @@ def new
...
     end
   end
 
   # TODO: This needs to be re-written.
   # If you are logged in and not an admin, this should fail.
   def create
     # Find by email, but allow updating other info.
-    @teacher = Teacher.find_by(email: teacher_params[:email])
+    teacher_id = EmailAddress.find_by(email: teacher_params[:primary_email])&.teacher_id
+    @teacher = Teacher.find_by(id: teacher_id)
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:56]: If you don't use `_id` then you can get the association record directly:

```suggestion
    @teacher = EmailAddress.find_by(email: teacher_params[:primary_email])&.teacher
```

File: app/controllers/teachers_controller.rb [hunk 10]
```diff
@@ -42,22 +42,25 @@ def new
...
     if @teacher && defined?(current_user.id) && (current_user.id == @teacher.id)
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:57]: This isn't for this PR, but the creation workflow when signed in seems like it doesn't make sense anymore. We should revisit this...

Comment (perryzjc) [reply app/controllers/teachers_controller.rb:57]: Marked. We will discuss it during next available meeting.

File: app/controllers/teachers_controller.rb (hunk 4 above)
Comment (perryzjc) [reply app/controllers/teachers_controller.rb:286]: My existing `@teacher.primary_email` is a method call to only get a email address value itself:
```ruby
  def primary_email
    # ||:email this code is temporary for this PR: https://github.com/cs169/BJC-Teacher-Tracker-App/pull/49
    # to make sure at least original data in db still work and passed the existing tests
    self[:email] || email_addresses.find_by(primary: true)&.email
  end
```

To update the `primary` attribute for old email address record, based on your advice, I feel it could be
```ruby
teacher.email_addresses.find_by(primary: true)&.update(primary: false)
```

**Comment by perryzjc** [comment]
@cycomachead, I have just finished updating based on the review you provided as of Apr 11, 2:30 PM. The only uncertain part I have is regarding separating the updating of personal email to a different controller:
image

I am not sure what is expected for separating that, as updating personal email was already in the original code under `#teacher:create`. So far, what I have done relevant to this part of the review is refactoring the parameter handling to be something like this, helping to reduce the code complexity:
image
image
image
